ocaml/quicktest/qt_filter.ml

```diff
@@ -0,0 +1,200 @@
+          end
+      | None -> ())
+
+let cleanup () = Client.Client.Session.logout ~rpc:!A.rpc ~session_id:!session_id
+
+let wrap f =
+  init ();
+  Xapi_stdext_pervasives.Pervasiveext.finally
+    (fun () -> f (); finish ())
+    cleanup
+
+let conn tcs = for_each (fun (name, speed, test) -> [(name, speed, test !A.rpc !session_id)]) tcs
+
+module SR = struct
+  type srs = unit -> Qt.sr_info list
+
+  let get_sr_info rpc session_id sr =
+    let get_sr_features session_id sR =
+      (* Even though the SM backend may expose a VDI_CREATE capability attempts
+         to actually create a VDI will fail in (eg) the tools SR and any that
+         happen to be R/O NFS exports *)
+      let avoid_vdi_create session_id sr =
+        let other_config = Client.Client.SR.get_other_config rpc session_id sr in
+        let is_tools_sr = Client.Client.SR.get_is_tools_sr rpc session_id sr in
+        let is_iso_sr = Client.Client.SR.get_content_type rpc session_id sr = "iso" in
+        let special_key = "quicktest-no-VDI_CREATE" in
+        let is_marked = List.mem_assoc special_key other_config && List.assoc special_key other_config = "true" in
+        is_tools_sr || is_iso_sr || is_marked
+      in
+
+      let caps =
+        let ty = Client.Client.SR.get_type rpc session_id sR in
+        let sm = Client.Client.SM.get_all_records rpc session_id in
+        match List.filter (fun (_, r) -> r.API.sM_type = ty) sm with
+        | [] -> Alcotest.failf "Could not find SM plugin for SR type %s" ty
+        | [ _, plugin ] -> plugin.API.sM_capabilities
+        | _ :: _ -> Alcotest.failf "Multiple SM plugins found for SR type %s" ty
+      in
+      let avoid_vdi_create = avoid_vdi_create session_id sR in
+      let caps =
+        if avoid_vdi_create then
+          List.filter
+            (fun cap -> not (List.mem cap Sr_capabilities.[ vdi_create; vdi_delete ]))
+            caps
+        else
+          caps
+      in
+      let ops = Client.Client.SR.get_allowed_operations ~session_id ~rpc:rpc ~self:sR in
+      let ops =
+        if avoid_vdi_create then
+          List.filter
+            (fun cap -> not (List.mem cap [`vdi_create; `vdi_destroy]))
+            ops
+        else
+          ops
+      in
+      (ops, caps)
+    in
+
+    let allowed_operations, capabilities = get_sr_features session_id sr in
+    let open Qt in
+    { sr
+    ; allowed_operations
+    ; capabilities
+    }
+
+  let list_selected_srs rpc session_id =
+    Client.Client.SR.get_all rpc session_id
+    |> List.filter
+      (fun sr ->
+         Client.Client.SR.get_PBDs rpc session_id sr
+         |> List.exists
+           (fun pbd -> Client.Client.PBD.get_currently_attached rpc session_id pbd)
+      )
+    |> List.map (get_sr_info rpc session_id)
+
+  let only sr () = [get_sr_info !A.rpc !session_id sr]
+
+  let all_srs = lazy (list_selected_srs !A.rpc !session_id)
+
+  let all =
```

Comment:
Do we need `begin ... end` here?

Reply:
No - I left it here for clarity, so I don't have to remember the rules about how `if` expressions are parsed :)